### PR TITLE
PLT-7884 Query contracts by party addresses

### PR DESCRIPTION
**Pull request by jhbertra**
- [x] `GetHeaders` query rewritten using dynamic SQL
- [x] `partyAddresses` and `partyRoles` added to contract filter
- [x] `partyAddress` and `partyRole` query params added to REST API
- [x] Indexing by party address and role token added with non-destructive migration task run before main indexer.
- [x] Golden tests added for `GetHeaders`
- [x] Integration test updated for `GetHeaders`

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
        - [ ] Change made to Marlowe validator (@bwbush and @palas must be included as reviewers)
        - Review not required
        - [ ] Minor changes to non-critical code, documentation, nix derivations, configuration files, or scripts
        - [ ] Formatting, spelling, grammar, or reorganization
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-integration-tests/test/Language/Marlowe/Runtime/Integration/MarloweQuery.hs [hunk 1]
```diff
@@ -96,21 +107,87 @@ instance PaginatedQuery GetHeaders where
...
+              filterVariations defaultContractFilter{tagsSym}
+          , do
+              roleCurrenciesSym <- Set.toList $ Set.powerSet $ Set.fromList $ allRefs q
+              filterVariations defaultContractFilter{roleCurrenciesSym}
+          , do
+              partyAddressesSym <- Set.toList $ Set.powerSet $ Set.fromList [minBound .. maxBound]
+              filterVariations defaultContractFilter{partyAddressesSym}
+          , do
+              partyRolesSym <- Set.toList $ Set.powerSet $ Set.fromList $ allRefs q
+              filterVariations defaultContractFilter{partyRolesSym}
```
Comment (jhbertra) [inline marlowe-integration-tests/test/Language/Marlowe/Runtime/Integration/MarloweQuery.hs:142]: I had to change the method of enumerating the filters, as the nested power sets were making the test suite way too huge. Even this method produces ~10,000 test cases, but continuing with the nested power sets lead to >100,000 test cases.

File: marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Database/PostgreSQL/CommitBlocks.hs [hunk 2]
```diff
@@ -680,6 +699,25 @@ unzip13 =
...
+        , e : es
+        , f : fs
+        , g : gs
+        , h : hs
+        , i : is
+        , j : js
+        )
+    )
+    ([], [], [], [], [], [], [], [], [], [])
+
```
Comment (jhbertra) [inline marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Database/PostgreSQL/CommitBlocks.hs:720]:
```suggestion
```

File: marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Database/PostgreSQL/CommitBlocks.hs [hunk 3]
```diff
@@ -437,69 +452,72 @@ withdrawTxToWithdrawalTxInRows BlockHeader{..} MarloweWithdrawTransaction{..} =
...
-  , V.fromList withdrawalTxInCreateTxIxRows
-  , V.fromList invalidApplyTxTxIdRows
-  , V.fromList invalidApplyTxInputTxIdRows
-  , V.fromList invalidApplyTxInputTxIxRows
-  , V.fromList invalidApplyTxBlockIdRows
-  , V.fromList invalidApplyTxErrorRows
-  , V.fromList contractTxOutTagTagRows
-  , V.fromList contractTxOutTagTxIdRows
-  , V.fromList contractTxOutTagTxIxRows
+  ( concat parties
```
Comment (jhbertra) [inline marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Database/PostgreSQL/CommitBlocks.hs:457]: I had to run these queries separately - apparently GHC cannot handle tuples larger than 74 items.

Comment (bwbush) [reply marlowe-runtime/indexer/Language/Marlowe/Runtime/Indexer/Database/PostgreSQL/CommitBlocks.hs:457]: 🤯

**Review by bwbush — APPROVED** [review]
Excellent new capability! Looks good.

I manually tested by . . . 
1. Ran on existing database and verified that the new tables were populated.
2. Ran from genesis and similarly verified.
3. Did a couple of ad-hoc queries via the REST API.

**Review by paluh — APPROVED** [review]
I don't want to block the PR. I really want this feature to be on main. I sketched quickly and of course it looks good :-)

**Pull request merged**